lib/python/treadmill_aws/autoscale.py

```diff
@@ -145,134 +148,188 @@ def delete_servers_by_name(servers):
+            continue
 
-    apps = response.json()
+        extra_servers.append(server['name'])
+    return extra_servers
 
-    admin_srv = admin.Server(context.GLOBAL.ldap.conn)
 
-    running = 0
-    pending = 0
-    busy_servers = set()
-    for app in apps:
-        if app['host']:
-            running += 1
-            busy_servers.add(app['host'])
-        else:
-            pending += 1
-
-    _LOGGER.info('Apps: running: %s, pending: %s', running, pending)
-
-    servers = admin_srv.list({'cell': context.GLOBAL.cell})
-    all_servers = {s['_id'] for s in servers}
-
-    idle_servers = all_servers - busy_servers
-    _LOGGER.info(
-        'Servers: busy: %s, idle: %s',
-        len(busy_servers),
-        len(idle_servers)
-    )
+def _scale_partition(app_server_ratio, autoscale_conf, apps, servers):
+    min_servers = autoscale_conf['min_servers']
+    max_servers = autoscale_conf['max_servers']
+    app_server_ratio = autoscale_conf.get('app_server_ratio', app_server_ratio)
 
-    return _STATE(
-        running=running,
-        pending=pending,
-        busy_srv_cnt=len(busy_servers),
-        idle_servers=list(idle_servers)
-    )
+    _LOGGER.debug('Apps: %r', apps)
+    _LOGGER.debug('Servers: %r', servers)
 
+    # Scaling rules:
+    #
+    # - If there are pending apps, add servers unless we have enough idle ones.
+    #   Servers needed = ceil(pending apps * ratio) - idle_servers.
+    #   Notice invariant: ratio = busy_servers / running_apps <= 1.
+    # - If there are no pending apps, shutdown idle servers (skip new servers).
+    # - If there are pending apps and we have enough idle servers, shutdown
+    #   (idle servers - pending apps) servers (skip new servers).
+    # - Shutdown down and frozen servers that have no apps placed on them.
+    # - Server count can't exceed max_servers or be less than min_servers.
+    #   Only new and up servers count towards min_serves, down or frozen don't.
 
-def _select_idle_servers(idle_servers, pending):
-    """Remove idle servers.
-    """
-    _LOGGER.info(
-        'Idle servers cleanup - idle: %s, pending: %s',
-        idle_servers,
-        pending
-    )
+    pending_apps = len([app for app in apps if not app['server']])
+    running_apps = 0
+    busy_servers = 0
+    idle_servers = 0
 
-    extra = []
-    can_remove = len(idle_servers) - pending
-    if can_remove > 0:
-        _LOGGER.info('Can delete idle servers: %s', can_remove)
-        extra = idle_servers[0:can_remove]
+    for server in servers:
+        if server['state'] in ('new', 'up'):
+            if server['num_apps']:
+                busy_servers += 1
+                running_apps += server['num_apps']
+            else:
+                idle_servers += 1
+
+    new_servers = 0
+    max_new_servers = max(0, max_servers - len(servers))
+
+    if running_apps > 0:
+        ratio = float(busy_servers) / float(running_apps)
+    else:
+        ratio = app_server_ratio
```

Comment:
nit - this is server_app_ratio, not app_server_ratio, right? should be renamed for clarity?

Reply:
Renamed.